### PR TITLE
SNOW-1658905 Complete Iceberg validations

### PR DESCRIPTION
**Pull request by sfc-gh-mbobowski**
# Overview

SNOW-1658905

Completes iceberg validations by implementing:
- record_metadata data type validation
- plug-in validations into the rest of code
- refactor to run some tasks once per topic
- refactor of calling `describe table` statement
- fix missing $ in bash script

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 1]
```diff
@@ -337,6 +317,7 @@ public boolean pipeExist(final String pipeName) {
   }
 
   @Override
+  // TODO - use describeTable()
```
Comment (sfc-gh-mbobowski) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:320]: Not in scope of this PR.

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 2]
```diff
@@ -398,6 +399,22 @@ static boolean isSnowpipeIngestion(Map<String, String> config) {
             .equalsIgnoreCase(IngestionMethodConfig.SNOWPIPE.toString());
   }
 
+  /**
+   * @param config config with applied default values
+   * @return true when Iceberg mode is enabled.
+   */
+  public static boolean isIcebergIngestion(Map<String, String> config) {
```
Comment (sfc-gh-akowalczyk) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:406]: should be rather `isIcebergIngestionEnabled`? :D

File: src/main/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergTableSchemaValidator.java [hunk 3]
```diff
@@ -15,19 +20,28 @@ public IcebergTableSchemaValidator(SnowflakeConnectionService snowflakeConnectio
...
+            .orElseThrow(() -> SnowflakeErrors.ERROR_0032.getException("table_not_found"));
+
+    DescribeTableRow metadata =
+        columns.stream()
+            .filter(c -> Objects.equals(c.getColumn(), TABLE_COLUMN_METADATA))
+            .findFirst()
+            .orElseThrow(
+                () -> SnowflakeErrors.ERROR_0032.getException("record_metadata_not_found"));
+
+    if (!metadata.getType().startsWith("OBJECT")) {
```
Comment (sfc-gh-akowalczyk) [inline src/main/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergTableSchemaValidator.java:39]: Why did you use `OBJECT` here? Could you please extract it to a separate variable with a name explaining the meaning of its value?

**Review by sfc-gh-akowalczyk — APPROVED** [review]
2 tiny things, besides that, it looks good 👍

File: src/main/java/com/snowflake/kafka/connector/internal/DescribeTableRow.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.snowflake.kafka.connector.internal;
+
+/** Class representing a single row returned by describe table statement. */
+public class DescribeTableRow {
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/DescribeTableRow.java:4]: nice

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 5]
```diff
@@ -1109,6 +1091,38 @@ public ChannelMigrateOffsetTokenResponseDTO migrateStreamingChannelOffsetToken(
...
+    PreparedStatement stmt = null;
+    List<DescribeTableRow> rows = new ArrayList<>();
+
+    try {
+      stmt = conn.prepareStatement(query);
+      stmt.setString(1, tableName);
+      ResultSet result = stmt.executeQuery();
+
+      while (result.next()) {
+        String columnName = result.getString(1);
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:1107]: we could also use names of the columns, bit safer but i guess the order won;t change in the future
```suggestion
        String columnName = result.getString("name");
```

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java [hunk 6]
```diff
@@ -1109,6 +1091,38 @@ public ChannelMigrateOffsetTokenResponseDTO migrateStreamingChannelOffsetToken(
...
+        String type = result.getString(2);
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:1108]: nit: as above
```suggestion
        String type = result.getString("type");
```

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java (hunk 5 above)
Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionServiceV1.java:1107]: Good point, it is not likely but might break the whole connector some day.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 7]
```diff
@@ -242,25 +246,25 @@ public void startPartition(String tableName, TopicPartition topicPartition) {
   @Override
   public void startPartitions(
       Collection<TopicPartition> partitions, Map<String, String> topic2Table) {
+    partitions.stream()
+        .map(TopicPartition::topic)
+        .distinct()
+        .forEach(topic -> perTopicActionsOnStartPartitions(topic, topic2Table));
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:252]: let's hope that it's gonna improve also rebalancing, as we discussed

File: src/main/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergTableSchemaValidator.java (hunk 3 above)
Comment (sfc-gh-wtrefon) [reply src/main/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergTableSchemaValidator.java:39]: this is the type that;s gonna be required for record_metadata column for iceberg.

But that's a good idea to have a class with iceberg types but not necessary in this PR, we can refactor it later

Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergTableSchemaValidator.java:39]: `OBJECT` is a Snowflake type (like `OBJECT()` or `OBJECT(id VARCHAR, age NUMBER)`. I will extract it to `SF_STRUCTURED_OBJECT` to be more precise on that.

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 2 above)
Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:406]: I will go with simpler `isIcebergEnabled()`

**Review by sfc-gh-wtrefon — APPROVED** [review]

**Pull request merged**